plugins/mana/plugin.go

```diff
@@ -371,8 +408,253 @@ func GetPendingMana(value float64, n time.Duration) float64 {
+	baseManaVector, err := mana.NewBaseManaVector(mana.ConsensusMana)
+	if err != nil {
+		return nil, nil, err
+	}
+	cbmvPast := baseManaVector.(*mana.ConsensusBaseManaVector)
+	cachedObj := consensusBaseManaPastVectorMetadataStorage.Get([]byte(mana.ConsensusBaseManaPastVectorMetadataStorageKey))
+	cachedMetadata := &mana.CachedConsensusBasePastManaVectorMetadata{CachedObject: cachedObj}
+	defer cachedMetadata.Release()
+
+	if cachedMetadata.Exists() {
+		metadata := cachedMetadata.Unwrap()
+		if t.After(metadata.Timestamp) {
+			consensusBaseManaPastVectorStorage.ForEach(func(key []byte, cachedObject objectstorage.CachedObject) bool {
+				cachedPbm := &mana.CachedPersistableBaseMana{CachedObject: cachedObject}
+				defer cachedPbm.Release()
+				p := cachedPbm.Unwrap()
+				err = cbmvPast.FromPersistable(p)
+				if err != nil {
+					log.Errorf("error while restoring %s mana vector from storage: %w", mana.ConsensusMana.String(), err)
+					baseManaVector, _ := mana.NewBaseManaVector(mana.ConsensusMana)
+					cbmvPast = baseManaVector.(*mana.ConsensusBaseManaVector)
+					return false
+				}
+				return true
+			})
+		}
+	}
+
+	var eventLogs []mana.Event
+	consensusEventsLogStorage.ForEach(func(key []byte, cachedObject objectstorage.CachedObject) bool {
+		cachedPe := &mana.CachedPersistableEvent{CachedObject: cachedObject}
+		defer cachedPe.Release()
+		pe := cachedPe.Unwrap()
+		if pe.Time.After(t) {
+			return true
+		}
+
+		// already consumed in stored base mana vector.
+		if cachedMetadata.Exists() && cbmvPast.Size() > 0 {
+			metadata := cachedMetadata.Unwrap()
+			if pe.Time.Before(metadata.Timestamp) {
+				return true
+			}
+		}
+
+		var ev mana.Event
+		ev, err = mana.FromPersistableEvent(pe)
+		if err != nil {
+			return false
+		}
+		eventLogs = append(eventLogs, ev)
+		return true
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+	sort.Slice(eventLogs, func(i, j int) bool {
+		var timeI, timeJ time.Time
+		var typeI, _ byte
+		switch eventLogs[i].Type() {
+		case mana.EventTypePledge:
+			timeI = eventLogs[i].(*mana.PledgedEvent).Time
+			typeI = mana.EventTypePledge
+		case mana.EventTypeRevoke:
+			timeI = eventLogs[i].(*mana.RevokedEvent).Time
+			typeI = mana.EventTypeRevoke
+		}
+
+		switch eventLogs[j].Type() {
+		case mana.EventTypePledge:
+			timeJ = eventLogs[j].(*mana.PledgedEvent).Time
+			_ = mana.EventTypePledge
+		case mana.EventTypeRevoke:
+			timeJ = eventLogs[j].(*mana.RevokedEvent).Time
+			_ = mana.EventTypeRevoke
+		}
+
+		if !timeI.Equal(timeJ) {
+			return timeI.Before(timeJ)
+		}
+
+		return typeI == mana.EventTypeRevoke
+	})
+	err = cbmvPast.BuildPastBaseVector(eventLogs, t)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	err = cbmvPast.UpdateAll(t)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return cbmvPast, eventLogs, nil
+}
+
+func checkConsensusEventStorage() {
+	size := consensusEventsLogStorage.GetSize()
+	if size < maxConsensusEventsInStorage {
+		return
```

Comment:
When we reach this number, the function body runs and creates a `ConsensusBaseManaVectorPast` at `now`, and prunes all the events up to this point. So If I wanted to ask for the vector at `now-t`, I would fail, as there are no more events in the log. Of course I could still go to the csv file.
But my problem with this approach is that it doesn't leave me with a buffer of time in the past which would let me get the vector after the pruning has happened.
How about having a sliding window instead?

![mana_event_storage](https://user-images.githubusercontent.com/32927267/104212135-2a067c00-5435-11eb-898d-329d296dfe37.png)

Instead of consuming all events up until `now` when building the `PastConsensusBaseManaVector`, consume only `SlidingInterval` number of events.
- How to determine then what `t` to feed into `GetPastConsensusManaVector(t)`? (or how to determine `t_newCBMV` on the figure): just like you do the chronological event sort within GetPastConsensusManaVector, you could get the `SlidingInterval` number oldest events from the event storage. `t_newCBMV` then will be the timestamp of the youngest among these.
- Once done, delete the consumed events only from the event storage.
- Wait for the number of events to reach `MaxEventInStorage` , and repeat.

Then we would be able to still query the `ConsensusBaseManaVector` in `now-t_newCBMV`.

P.S: when selecting the edge at the sliding interval, be careful to not separate events belonging to the same transaction (they will have the same timestamp anyway).

Reply:
good point